### DISCUSSION TITLE
ssr.noExternal option — any downsides?

### DISCUSSION THREAD

**Discussion by plashenkov** [frandiox/vite-ssr#112] [Q&A] [answered]
I'm thinking about using the `ssr.noExternal` Vite option. I can see that the `package.json` remains more clean: you have only `express` in `dependencies`, and all other libraries go to `devDependencies`, and later they are not needed in production, everything is compiled into one `server/app.js` file. So, it looks nice.

Plus, later you do not have to think when installing yet another dependency — is this a dev-only dependency or maybe I need it  in production. So, another pro — convenience.

But are there any downsides? Any significant impact to a dev-server mode, build, and — the most important — a production mode. Does having external dependencies really provide advantages?

**Comment by frandiox** [comment, marked as answer]
It definitely makes things easier when you deploy to prod. However, Vite's bundling is not perfect (I think it's marked as experimental, together with all the SSR stuff). It has issues with some dependencies such as `aws-sdk`, etc. and this can change in any Vite update.

Other than that, as far as I understand, Vite would wrap your CommonJS modules in some extra code, making it slightly slower to load compared to a raw `require` in Node. But the runtime should be similar.

This wouldn't affect the dev server because Vite only bundles during build.

So I would say, try to bundle your app and if it works, go for it 👍 

-- If you have e2e/integration tests, I would run them against the built app rather than against a dev server so you make sure the bundling is not breaking anything.

**Comment by plashenkov** [comment]
Thanks! Very appreciate your answer!
